python/sanic/server.py

```diff
@@ -17,3 +21,11 @@ async def user_info(request, id):
 @app.route("/user", methods=["POST"])
 async def user(request):
     return text("")
+
+if __name__ == "__main__":
+    try:
+        workers = int(sys.argv[1])
+    except IndexError:
+        workers = multiprocessing.cpu_count()
+    app.run(host='0.0.0.0', port=3000, workers=workers,
+            debug=False, access_log=False)
```

Comment:
why disabling access logs ?

Reply:
According to my review, this is the way other frameworks were being benchmarked. In particular, I noted this across all of the ASGI frameworks (which all should benchmark fairly close to the same `70000` that Sanic does in ASGI mode), and also Flask (which is around 50k). I did not dig into the Django benchmark, that's a whole different beast.

And, I think that sort of makes sense. Writing to `stdout` in Python is terribly slow, and is not a good practice.

In my experience on Sanic projects, and in general, a **much** better solution would be to put `nginx` as a proxy in front of your server, and handle access logs there.

Indeed, adding `access_log=True` changes the results **considerably**.

```
Running 30s test @ http://127.0.0.1:3000
  12 threads and 400 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    13.17ms   12.34ms 157.35ms   83.24%
    Req/Sec     2.99k     1.00k   11.33k    73.22%
  1071363 requests in 30.08s, 120.56MB read
Requests/sec:  35620.79
Transfer/sec:      4.01MB
```

More importantly, however, I think there needs to be some guidelines and transparency about this because at least for Python, that makes a **huge** difference, and is rarely implemented in production. I am a big advocate for allowing tools like `nginx` and `apache` do what they do best (including logging).

Reply:
> this is the way other frameworks were being benchmarked

Yes, but **this** is wrong

Correct me if I'm wrong, but access log should **NOT** be disabled. I have use some `python` **framework** in **production**, but _never_ disabled logs

Reply:
This is your benchmark, so it is your call. But, to me this project is useful to see how these frameworks compare with one another in an attempt at objectivity. Access logs have no use here for several reasons:

1. It is an absurd degradation [to just write to `stdout`](https://stackoverflow.com/questions/3857052/why-is-printing-to-stdout-so-slow-can-it-be-sped-up). This should **not** be done in production. It is terrible for performance, yes, but it is also drain on system resources because of the lack of buffering. I don't know many production systems logging like this. If they are, they shouldn't be.
1. It clouds the evaluations of the frameworks.
1. If you wanted a "production" ready implementation, I would highly suggest running some builds inside containers so that there would be multiple potential layers, load balancing, etc. To me this is much too complicated and brings the utility of this benchmarking down since it vastly increases the variables involved. 
1. If you were really looking for a production ready build, the apps need to be much more complicated. There are **VAST** differences that can be introduced [based upon the complexity of the routing](https://github.com/vltr/very-stupid-simple-routers-bench). Your tests do not account for this, or other network calls, which therefore does not approximate production usage. This is **NOT** a fault. I am not critiquing. I think the simplistic approach makes sense for what you are trying to achieve. With that in mind, using sub-optimal, non-production worthy logging sort of breaks that mold.

Reply:
why are you saying that log are just writing to _stdout_ ? I don't know the internal of `gunicorn` (for example), so I can't say.

However, when you deploy an application in **production**, it is a `non-sens` for me to run an application in **production** without logging (not speaking about logging level).

For me this benchmark is useful to see how technologies can react (performances, but other insights will come) in the real-word.

I agree with your 4 points, but :
1. I mean logging, no matter if it is `stdout` (generally it is not)
2. Yes, writing to `stdout` could mess result, but why writing to `stdout` ?
3. In reflection ... it's very complex, but totally feasible (I admin, I'm an optimistic guy :heart: )
4. In reflection ... this project will not be only a benchmark, but also a tool benchmarking a more complex implementation (not made by me, perhaps https://github.com/gothinkster/realworld)

Finally, I had to say that there is plenty of benchmark all over the internet. The very value of this project is to (mainly) reply to 2 cases :
+ I have to choose a stack for a new project
+ I want to improve my framework

Reply:
Like I said, your project, your goals. I am not here to tell you how to do it. Just merely pointing out that if you want to have a fair shake at representation of various strategies, then you need **much** more complex setups for each project, which probably includes involving many more maintainers. Having looked through your setups (for the Python frameworks at least), they are still far too simplistic to be realistic.

If you are going for "realism", why not have more complex apps with more routes? I think this would be very helpful to see how frameworks handle complex routing, which in my experience has a major impact on performance.

Also, any change that you could set the project up to handle containers? The reason I ask is that I think the **best** (in an overly generalized way) and most practical deployment strategy for any of the Python frameworks would be to sit them behind `nginx`.

Reply:
Also, if the goal is to help people decide what to choose, shouldn't you also have some factor for showing popularity, download rates, code update frequency, etc?

Reply:
> If you are going for "realism", why not have more complex apps with more routes?

Think of it ...

>  sit them behind nginx

I'll also produce some results, with nginx alongside bjoern ...

> showing popularity, download rates, code update frequency, etc?

Yes, in a second time. I have to keep focus on results right now

Could logging be enable for sanic ? It use stdout or logging python module ?

Reply:
Standard python module. By default the config is set to output to st out. But I will push another update to use the null logger so it is doing the logging.

Reply:
I spent some more time looking through this... I really think that access logs have no place here. Most of the other projects that you are testing do **not** even offer them, and from what I can see, you have everything set to `WARNING`. Which means that there would not even be any access logs.

I would suggest merging this PR with a similar logging config as to the other projects, and then coming back to this question afterwards and doing a single PR to add consistent logging across all tests. This seems to be the best way to get "fair" results. Right now Sanic is listed at the bottom of the Python frameworks, and it should be listed as one of the faster Python frameworks.

Reply:
> By default the config is set to output to st out.

I has not aware of that. How, then log are used in **production** (`rsyslog` or an other strategy) ?

> I would suggest merging this PR with a similar logging config as to the other projects, and then coming back to this question afterwards and doing a single PR to add consistent logging across all tests. 

Absolutely, the logging issue **SHOULD** not block this `PR` :heart:

Reply:
> How, then log are used in production (rsyslog or an other strategy) ?

I can only speak anecdotally since I have never done a survey of the issue. However, I suspect the most normal setup is to use `nginx` in front and log to syslog or something like that. Perhaps I am mistaken.

But, the Sanic docs do sort of say that the access logging that is enabled by default is not performant. It is there to help users get up and running fast. The goal of the project is not only performance, but also fast development.

Reply:
Ok. That's relevant. We can then investigate / decide in https://github.com/the-benchmarker/web-frameworks/issues/2622